api/net/ip4/icmp4.hpp

```diff
@@ -0,0 +1,210 @@
+     */
+    void destination_unreachable(Packet_ptr pckt, icmp4::code::Dest_unreachable code);
+
+    /**
+     *
+     */
+    void redirect(icmp4::Packet& req, icmp4::code::Redirect code);
+
+    /**
+     *  Sending a Time Exceeded message from a host when fragment reassembly time exceeded (code 1)
+     *  Sending a Time Exceeded message from a gateway when time to live exceeded in transit (code 0)
+     */
+    void time_exceeded(Packet_ptr pckt, icmp4::code::Time_exceeded code);
+
+    /**
+     *  Sending a Parameter Problem message if the gateway or host processing a datagram finds a problem with
+     *  the header parameters such that it cannot complete processing the datagram. The message is only sent if
+     *  the error caused the datagram to be discarded
+     *  Code 0 means Pointer (uint8_t after checksum) indicates the error/identifies the octet where an error was detected
+     *  in the IP header
+     *  Code 1 means that a required option is missing
+     */
+    void parameter_problem(Packet_ptr pckt, uint8_t error);
+
+    // May
+    void timestamp_request(IP4::addr ip);
+    void timestamp_reply(icmp4::Packet& req);
+
+    void ping(IP4::addr ip);
+    void ping(IP4::addr ip, icmp_func callback);
+
+  private:
+    static int request_id_; // message identifier for messages originating from IncludeOS
+    Stack& inet_;
+    downstream network_layer_out_ =   nullptr;
+    uint8_t includeos_payload_[48] =  {'I','N','C','L','U','D',
+                                      'E','O','S',1,2,3,4,5,
+                                      'A','B','C','D','E','F','G','H',
+                                      'I','J','K','L','M','N','O','P',
+                                      'Q','R','S','T','U','V','W','X',
+                                      'Y','Z',1,2,3,4,5,6,
+                                      7,8};
+
+    struct ICMP_callback {
+      using icmp_func = ICMPv4::icmp_func;
+      using Tuple = ICMPv4::Tuple;
+
+      Tuple tuple;
+      icmp_func callback;
+      Timers::id_t timer_id;
+
+      ICMP_callback(ICMPv4& icmp, Tuple t, icmp_func cb)
+      : tuple{t}, callback{cb}
+      {
+        timer_id = Timers::oneshot(std::chrono::seconds(40), [&icmp, t](Timers::id_t) {
+          icmp.remove_ping_callback(t);
+        });
+      }
+    };  // < struct ICMP_callback
+
+    std::map<Tuple, ICMP_callback> ping_callbacks_;
+
+    void forward_to_transport_layer(icmp4::Packet& req);
+
+    /**
+     *  Find the ping-callback that this packet is a response to, execute it and erase the object
+     *  from the ping_callbacks_ map
+     */
+    inline void execute_ping_callback(icmp4::Packet& ping_response) {
+      // Find callback matching the reply
+      auto it = ping_callbacks_.find(std::make_pair(ping_response.id(), ping_response.sequence()));
+
+      if (it != ping_callbacks_.end()) {
+        it->second.callback(ICMP_packet{ping_response.id(), ping_response.sequence(), ping_response.ip().ip_src(),
```

Comment:
To make it cleaner you could add a constructor to `struct ICMP_packet` that takes a `const icmp4::Packet&` and hide all these lines there (with a shorter name).